src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataSymbolMap.cs

```diff
@@ -0,0 +1,131 @@
+            ImmutableDictionary<ITypeSymbol, TInfo>.Builder interfaceInfosBuilder = ImmutableDictionary.CreateBuilder<ITypeSymbol, TInfo>();
+
+            foreach (TInfo info in taintedDataInfos)
+            {
+                if (wellKnownTypeProvider.TryGetTypeByMetadataName(info.FullTypeName, out INamedTypeSymbol namedTypeSymbol))
+                {
+                    if (info.IsInterface)
+                    {
+                        interfaceInfosBuilder[namedTypeSymbol] = info;
+                    }
+                    else
+                    {
+                        concreteInfosBuilder[namedTypeSymbol] = info;
+                    }
+                }
+            }
+
+            this.ConcreteInfos = concreteInfosBuilder.ToImmutable();
+            this.InterfaceInfos = interfaceInfosBuilder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Mapping for concrete types.
+        /// </summary>
+        private ImmutableDictionary<ITypeSymbol, TInfo> ConcreteInfos { get; }
+
+        /// <summary>
+        /// Mapping for interface types.
+        /// </summary>
+        private ImmutableDictionary<ITypeSymbol, TInfo> InterfaceInfos { get; }
+
+        /// <summary>
+        /// Indicates that this mapping is empty, i.e. there are no types referenced by the compilation represented by the <see cref="WellKnownTypeProvider"/>.
+        /// </summary>
+        public bool IsEmpty { get { return this.ConcreteInfos.IsEmpty && this.InterfaceInfos.IsEmpty; } }
+
+        /// <summary>
+        /// Gets an enumeration of infos for the given type.
+        /// </summary>
+        /// <param name="namedTypeSymbol">Type to find infos for.</param>
+        /// <returns>Relevant infos for the given type.</returns>
+        public IEnumerable<TInfo> GetInfosForType(INamedTypeSymbol namedTypeSymbol)
+        {
+            Debug.Assert(namedTypeSymbol != null);
+
+            if (namedTypeSymbol == null)
+            {
+                yield break;
+            }
+
+            if (!this.InterfaceInfos.IsEmpty)
+            {
+                if (namedTypeSymbol.TypeKind == TypeKind.Interface
+                    && this.InterfaceInfos.TryGetValue(namedTypeSymbol, out TInfo infoForInterfaceSymbol))
+                {
+                    yield return infoForInterfaceSymbol;
+                }
+
+                foreach (INamedTypeSymbol interfaceSymbol in namedTypeSymbol.AllInterfaces)
+                {
+                    if (this.InterfaceInfos.TryGetValue(interfaceSymbol, out TInfo info))
+                    {
+                         yield return info;
+                    }
+                }
+            }
+
+            if (!this.ConcreteInfos.IsEmpty)
+            {
+                for (INamedTypeSymbol typeSymbol = namedTypeSymbol; typeSymbol != null; typeSymbol = typeSymbol.BaseType)
```

Comment:
>for (INamedTypeSymbol typeSymbol = namedTypeSymbol; typeSymbol != null; typeSymbol = typeSymbol.BaseType) [](start = 16, length = 105)

Nit: use [GetBaseTypesAndThis](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Unfactored/Roslyn/Core/ISymbolExtensions.cs#L24) helper and convert to foreach loop.